dashboard: Offer to add locked identity to keyring

Gundersanne opened the pull request:
release notes (not sure if this is release material but marking it anyway to decide later):


> When adding a machine to the dashboard fails because of a locked identity, cockpit will now offer an inline dialog to unlock this identity and add it to the ssh agent. 

---

![image](https://user-images.githubusercontent.com/11140201/70457675-d480d280-1ab0-11ea-87d2-c7c012cc71e4.png)

![image](https://user-images.githubusercontent.com/11140201/69881977-fc14b580-12ce-11ea-864b-84d3e8bbe460.png)

![image](https://user-images.githubusercontent.com/11140201/70457951-6ab4f880-1ab1-11ea-91cd-44498c610f50.png)

---

Gundersanne commented:
@garrett I want to make sure if this is ok design-wise

garrett commented:
Overall, I think the UI looks pretty good. It would just need a few changes here and there.

I'd suggest:
- Add more space between the "Cockpit was unable to..." and the user name
- "Available" should probably be something like "Key types"?
- Are the encryption key types all always id_***? Is that always an acronym? How many can there be?
  - Perhaps drop the "id_" part, uppercase the rest, and join them into an inline (comma-separated) list
- "Locked", and its passphrase-expecting form, should probably say something about expecting a passphrase.
- The "Log in" should probably be disabled when keys are locked. And in this case, perhaps make the "Unlock" button styled as a default button too, as it's the action that must be performed to proceed.

Gundersanne commented:
I changed the button style, added a little text, and increased the spacing. Note that most of this dialog was already in cockpit, I just added the `Locked` bit. 

Personally I think "available credentials" makes sense, or "available keys" makes sense as well. 

Also these aren't keytypes, the naming is just the filename of a key known to ssh-agent.

martinpitt commented on pkg/lib/machine-change-auth.html:
```diff
@@ -71,6 +71,21 @@ <h4 class="modal-title" translate="yes">Log in to {{host}}</h4>
+                <td class="top">
+                    <div class="key-locked">
+                        <div class="locked-identity"></div>
+                        <div translate="yes">Enter passphrase to add identity to ssh-agent</div>
```
So that means we don't even redirect the user to the existing "Authentication" dialog, but it's done inline in the Troubleshooting dialog? That's a nice workflow indeed! :+1 I'm sure a lot of users don't even know that the Auth dialog exists.

Perhaps this actually could just re-use the existing dialog somehow, but as the code is written now, let's use it.

Gundersanne replied:
Yea it uses the same `Keys` function from `pkg/lib/credentials.js`, but not the dialog itself.

martinpitt commented on pkg/lib/machine-dialogs.js:
```diff
@@ -640,6 +641,28 @@ function ChangeAuth(dialog) {
 
+        if (key_locked_div && error_options &&
+            error_options.error && error_options.error.startsWith("locked identity")) {
+            console.log("Locked identitiy: ", error_options);
```
it's "identititiy" of course! :wink:

martinpitt commented on pkg/lib/machine-dialogs.js:
```diff
@@ -640,6 +641,28 @@ function ChangeAuth(dialog) {
+            key_locked_div.find(".btn-primary").on("click", () => {
+                key_locked_div.find(".btn-primary").attr("disabled", true);
+                keys.load(identity_path, key_locked_div.find(".locked-identity-password").val())
+                        .done(() => dialog.render_error([]))
```
Passing `[]` here looks a little weird -- in the case of an exception it passes an ex object, not an array. This would look clearer with `null`? This gets passed to `cockpit.message()` as well, which expects an object, not an array.

Also, could this please use `.then()` and `.catch()`, instead of the deprecated jquery promise handlers?

Gundersanne replied:
This is a bit of a hack to reuse the error reporting within the dialog. Normally it waits until you click accept to clear the error. Maybe i should just add an explicit "clear_errors" function? Passing an empty object or null doesn't work, results in an error message being present.

martinpitt commented on pkg/lib/machine-dialogs.js:
```diff
@@ -640,6 +641,28 @@ function ChangeAuth(dialog) {
+
+            for (const id in keys.items) {
+                const key = keys.items[id];
+                if (key.name === identity_path.split("/").reverse()[0] && key.loaded)
```
`.pop()` instead of `.reverse()[0]`

martinpitt commented on test/verify/files/ssh/id_ed25519:
```diff
@@ -1,8 +1,8 @@
-TbTBigljz/ZoOReJtBO/2PHbtgm2ekM8UdTnB3VWjxLSH+z7UpgVHtqnuN4PQ+Rueq1kMq
-AoAamod8u1gb7RL9tcxGBRRrDMoGnO/f3QZIOEKDvqKSDy/PXB+Lmh4od1nmCQBhMsTMWk
-1hSB9/iF9UGn+mvw==
+b3BlbnNzaC1rZXktdjEAAAAACmFlczI1Ni1jdHIAAAAGYmNyeXB0AAAAGAAAABBxFawn02
```
How did that file get changed? Changing the password from "foobar" to "locked"? Why was that done, do we use the login password to try and unlock ssh keys? (that sounds a bit weird). The commit message should point that out.

Gundersanne replied:
Yea if the password is the same as the user password, it'll just automatically get added to the keyring. I'm not sure at what point that happens (for sure before cockpit-ssh gets invoked of course).

martinpitt commented on test/verify/check-multi-machine-key:
```diff
@@ -251,6 +252,42 @@ class TestMultiMachineKeyAuth(MachineCase):
+        b.wait_in_text("#dashboard_setup_server_dialog", "Fingerprint")
+        b.click("#dashboard_setup_server_dialog .btn-primary")
+        b.wait_in_text(".locked-identity", "/home/admin/.ssh/id_ed25519")
+        self.assertNotIn(".keys", "id_ed25519")
```
This is missing a `b.text()`

martinpitt commented on test/verify/check-multi-machine-key:
```diff
@@ -55,10 +55,9 @@ KEY_IDS = [
+    "256 b5:80:1b:f5:98:89:2a:39:f3:78:b3:64:5c:64:33:17 /home/admin/.ssh/id_ed25519 (ED25519)",
 ]
 
-
```
Please keep two empty lines between classes, that's PEP-8 style. Doesn't test-static-code complain?

martinpitt commented on test/verify/check-multi-machine-key:
```diff
@@ -251,6 +252,42 @@ class TestMultiMachineKeyAuth(MachineCase):
+        b.wait_in_text(".keys", "id_ed25519")
+        b.click("#dashboard_setup_server_dialog .modal-footer>.btn-primary")
+        b.wait_present("#dashboard-hosts .list-group-item[data-address='10.111.113.2']")
+        self.allow_hostkey_messages()
```
likewise, please keep two empty lines

martinpitt commented on pkg/shell/shell.less:
```diff
@@ -287,6 +287,10 @@ html.index-page body {
+#machine-change-auth > p {
+    margin-bottom: 5px;
+}
+
```
This should be squashed IMHO

Gundersanne commented on src/ssh/cockpitsshrelay.c:
```diff
@@ -977,14 +979,38 @@ do_key_identity_auth (CockpitSshData *data)
+  rc = ssh_pki_import_pubkey_file (pub_key_path, &pub_key);
+  if (rc == SSH_ERROR)
+    {
+      g_debug ("Error importing public key %s", pub_key_path);
```
`ssh_pki_import_pubkey_file` returns one of three values: `SSH_OK`, `SSH_ERROR`, or `SSH_EOF`.

`SSH_EOF` means the file didn't exist, or we don't have permission to read it. So in that case we should still try to use the private keyfile.

`SSH_ERROR` means something else, unspecified, went wrong, I guess we could still try to use private key in that case.

Gundersanne commented on src/ssh/cockpitsshrelay.c:
```diff
@@ -977,14 +979,38 @@ do_key_identity_auth (CockpitSshData *data)
+  if (rc == SSH_EOF)
+    {
+      g_debug ("Public key file %s doesn't exist or isn't readable", pub_key_path);
+    }
```
Oh indeed, there's `ssh_get_error` :)

garrett commented:
Does the inline alert "Password not accepted" look like it does in the screenshot above? It should look like a PF4 inline notification.

(I think @KKoukiou ported all of our existing ones to PF4 and that PR already landed in master?)

Gundersanne commented:
@garret Added a commit on top which swiches the alerts to PF4 (see screenshot too).

Gundersanne commented on pkg/lib/patterns.js:
```diff
@@ -48,10 +49,12 @@ function field_error(target, error) {
 
 function global_error(sel, error) {
-    var alert = $("<div class='alert alert-danger dialog-error'>");
+    var alert = $("<div class='pf-c-alert pf-m-danger pf-m-inline dialog-error' aria-label='inline danger alert'>");
```
I copied this from  https://github.com/cockpit-project/cockpit/pull/13038/files#diff-9703b1f6e42cd0c9b1395fa2df224e91

I can swap out the aria-label for the role though.

garrett replied:
I was thinking of a recent article I read (and I think I shared last Wednesday, perhaps?) about not explaining things in aria-label:
https://adrianroselli.com/2019/10/stop-giving-control-hints-to-screen-readers.html

I _think_ that applies here — especially if English isn't an understood language by the person using Cockpit with a screen reader (eventually). I believe `role="alert"` might be preferable. But I could also be wrong here.

We might also want to change the place you linked and do an aria-label audit across Cockpit at some point. (And at some point in time we should probably ask @jgiardino what the best practice is for stuff like this.)

As for the PR, I guess it's fine if we land it as-is, as we're using the same pattern elsewhere and aren't sure what's best — provided we open up and issue and revisit this item after the PR lands.

garrett replied:
I've opened an issue about this @ https://github.com/cockpit-project/cockpit/issues/13271 — so feel free to skip changing this on the PR; we can handle it later. (Ideally in January, after we're all back from holidays.)

Gundersanne replied:
Alright for consistency's sake I'll change it back to `aria-label` and then we can change it everywhere in one go.

Gundersanne commented on pkg/lib/machine-change-auth.html:
```diff
@@ -71,6 +71,21 @@ <h4 class="modal-title" translate="yes">Log in to {{host}}</h4>
+                    <div class="key-locked">
+                        <div class="locked-identity"></div>
+                        <div translate="yes">Enter passphrase to add identity to ssh-agent</div>
+                        <div>
```
This is with `<div class="key-locked form-group">`, without the div around the input/button. The alignment looks the same, but no empty div.

![image](https://user-images.githubusercontent.com/11140201/70538481-76142c80-1b62-11ea-8992-05165dd708dd.png)

Gundersanne commented on src/ssh/cockpitsshrelay.c:
```diff
@@ -977,14 +979,37 @@ do_key_identity_auth (CockpitSshData *data)
+  else if (rc == SSH_ERROR)
+    {
+      msg = ssh_get_error (data->session);
+      g_debug ("Error importing public key %s: error", pub_key_path, msg);
```
ack

Gundersanne commented on src/ssh/cockpitsshrelay.c:
```diff
@@ -977,14 +979,37 @@ do_key_identity_auth (CockpitSshData *data)
+    }
+  else if (rc == SSH_EOF)
+    {
+      g_debug ("Public key file %s doesn't exist or isn't readable", pub_key_path);
```
Exactly, this isn't bad at all.
